src/Stack/Lock.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Stack.Lock where
+
+import qualified Data.Yaml as Yaml
+import Data.Yaml (object)
+import Path (addFileExtension, fromAbsFile)
+import Path.IO (doesFileExist, getModificationTime)
+import qualified RIO.ByteString as B
+import Stack.Prelude
+import Stack.Types.Config
+
+data LockException =
+    LockNoProject
+    deriving (Typeable)
+
+instance Exception LockException
+
+instance Show LockException where
+    show (LockNoProject) = "No project found for locking."
+
+generateLockFile :: Path Abs File -> RIO Config ()
+generateLockFile stackFile = do
+    logDebug "Gennerating lock file"
+    mproject <- view $ configL . to configMaybeProject
+    p <-
+        case mproject of
+            Just (p, _) -> return p
+            Nothing -> throwM LockNoProject
+    let deps :: [RawPackageLocation] = projectDependencies p
+        resolver :: RawSnapshotLocation = projectResolver p
+    lockFile <- liftIO $ addFileExtension "lock" stackFile
+    resolver' :: SnapshotLocation <- completeSnapshotLocation resolver
+    deps' :: [PackageLocation] <- mapM completePackageLocation' deps
+    let deps'' = zip deps deps'
+    let depsObject =
+            Yaml.object
+                [ ( "resolver"
+                  , object
+                        [ ("original", Yaml.toJSON resolver)
+                        , ("complete", Yaml.toJSON resolver')
+                        ])
+                , ( "dependencies"
+                  , Yaml.array
+                        (map (\(raw, comp) ->
+                                  object
+                                      [ ("original", Yaml.toJSON raw)
+                                      , ("complete", Yaml.toJSON comp)
+                                      ])
+                             deps''))
+                ]
+    B.writeFile (fromAbsFile lockFile) (Yaml.encode depsObject)
+
+hasLockFile :: HasEnvConfig env => BuildConfig -> RIO env Bool
+hasLockFile bconfig = do
+    let stackFile = bcStackYaml bconfig
+    lockFile <- liftIO $ addFileExtension "lock" stackFile
+    liftIO $ doesFileExist lockFile
+
+-- parsePLI :: HasEnvConfig env => BuildConfig -> RIO env [PackageLocation]
+-- parsePLI bconfig = do
+--     let stackFile = bcStackYaml bconfig
+--         rootDir = parent stackFile
+--     lockFile <- liftIO $ addFileExtension "lock" stackFile
+--     (pli :: Yaml.Value) <- Yaml.decodeFileThrow (toFilePath lockFile)
+--     plis <- Yaml.parseMonad (parseLockFile rootDir) pli
+--     plis' <- liftIO $ plis
+--     pure $ NE.toList plis'
+isLockFileOutdated :: Path Abs File -> RIO Config Bool
```

Comment:
It's just an optimization step. Any reason you feel that this won't work ?

Reply:
it should work most of the times but there's also a quite easy way to get into an inconsistent state - just use an `stack.yaml` which is older that a lock file (e.g. extracted from an archive or received some other way which gives you not the current date/time in the end). Lock file is supposed to be small so I think it's better to read it every time so we could exclude such a chance of inconsistency.

Reply:
@qrilka I have just explained how even in the presence of an older `stack.yaml` file this will work here: https://github.com/commercialhaskell/stack/pull/4550#discussion_r257611144